src/dd/Approximation.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+ * Copyright (c) 2025 Munich Quantum Software Company GmbH
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+#include "dd/Approximation.hpp"
+
+#include "dd/Complex.hpp"
+#include "dd/ComplexNumbers.hpp"
+#include "dd/DDDefinitions.hpp"
+#include "dd/Node.hpp"
+#include "dd/Package.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <forward_list>
+#include <limits>
+#include <queue>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+namespace dd {
+namespace {
+constexpr uint16_t FLAG_DELETE = 0b100;
+constexpr uint16_t FLAG_LEFT = 0b010;
+constexpr uint16_t FLAG_RIGHT = 0b001;
+
+/**
+ * @brief A node-contribution pair and distance value for prioritization.
+ */
+class LayerNode {
+public:
+  explicit LayerNode(vNode* node)
+      : ptr(node), contribution(1.),
+        distance(std::numeric_limits<double>::max()) {}
+
+  LayerNode(vNode* node, double contribution, double budget)
+      : ptr(node), contribution(contribution),
+        distance(std::max<double>(0, contribution - budget)) {}
+
+  bool operator<(const LayerNode& other) const {
+    if (distance == 0 && other.distance == 0) {
+      return contribution < other.contribution;
+    }
+    return distance > other.distance;
+  }
+
+  vNode* ptr;
+  double contribution;
+
+private:
+  double distance;
+};
+
+/// @brief Priority queue of nodes, sorted by their distance to a given budget.
+using Layer = std::priority_queue<LayerNode>;
+/// @brief Maps nodes to their respective contribution.
+using Contributions = std::unordered_map<vNode*, double>;
+/// @brief Maps old nodes to rebuilt edges.
+using Lookup = std::unordered_map<const vNode*, vEdge>;
+/// @brief A node, a flag indicating left or right, and its contribution.
+using Terminal = std::tuple<vNode*, uint16_t, double>;
+
+/**
+ * @brief Search and mark nodes for deletion until the budget 1 - @p fidelity is
+ * exhausted.
+ * @details Uses a prioritized iterative-deepening search.
+ * Iterating layer by layer ensures that each node is only visited once.
+ */
+ApproximationMetadata mark(VectorDD& state, const double fidelity) {
+  Layer curr{};
+  curr.emplace(state.p);
+
+  std::forward_list<Terminal> candidates{};
+
+  ApproximationMetadata meta{fidelity, 0, std::numeric_limits<Qubit>::max()};
+
+  double budget = 1 - fidelity;
+  while (budget > 0) {
+    Contributions c; // Stores contributions of the next layer.
```

Comment:
Just a thought: might it be more performant to allocate this hash map once and then to simply clear it in each iteration of the whole loop?
That might safe some memory allocations and resizing operations on subsequent iterations.